File: panel/server/src/download.ts
import { createReadStream } from 'node:fs'
import { lstat, realpath, stat } from 'node:fs/promises'
import { PassThrough, Readable } from 'node:stream'
import { basename, join, resolve, sep } from 'node:path'
import type { Context } from 'hono'
import archiver from 'archiver'
import type { PanelConfig } from './panel-config.js'
import { Limiter } from './limit.js'
import { MANIFEST_FILE } from '@edge-sync/protocol'

// 版本目录内元数据文件名与内核 engine.ManifestFile 对齐。

const VERSION_RE = /^\d{8}-\d{6}(-\d+)?$/
const limiter = new Limiter(2)

/** 版本目录解析：current → realpath；版本名校验防穿越。 */
export async function resolveVersionDir(cfg: PanelConfig, task: string, version: string): Promise<string> {
  if (!task || task.includes('/') || task.includes('..')) {
    throw new HttpError(400, 'invalid task name')
  }
  const dataDir = join(resolve(cfg.dataDir), task)
  if (version === '' || version === 'current') {
    const link = join(dataDir, 'current')
    try {
      return await realpath(link)
    } catch {
      throw new HttpError(404, 'no current version yet')
    }
  }
  if (!VERSION_RE.test(version)) {
    throw new HttpError(400, `invalid version name: ${version}`)
  }
  const dir = join(dataDir, 'versions', version)
  const fi = await stat(dir).catch(() => null)
  if (!fi?.isDirectory()) {
    throw new HttpError(404, `version not found: ${version}`)
  }
  return await realpath(dir)
}

/** 文件解析：SafeJoin 语义 + 必须是常规文件；解析结果必须仍在版本目录内。 */
export async function resolveFileUnder(versionDir: string, subPath: string): Promise<string> {
  if (!subPath || subPath.includes('\0')) {
    throw new HttpError(400, 'invalid path')
  }
  const cleaned = resolve(versionDir, subPath)
  const versionReal = await realpath(versionDir)
  const candidate = await realpath(cleaned).catch(() => null)
  if (!candidate || !(candidate + sep).startsWith(versionReal + sep)) {
    throw new HttpError(400, 'path escapes version directory')
  }
  const fi = await lstat(candidate)
  if (!fi.isFile()) {
    throw new HttpError(400, 'not a regular file')
  }
  return candidate
}

export class HttpError extends Error {
  constructor(
    readonly status: number,
    message: string,
  ) {
    super(message)
  }
}

export function httpErrorOf(e: unknown): number {
  if (e instanceof HttpError) return e.status
  return 500
}

function contentDisposition(filename: string): string {
  return `attachment; filename*=UTF-8''${encodeURIComponent(filename)}`
}

/** GET /api/tasks/:name/download?version=&path= —— fs 流式单文件下载。 */
export async function handleDownload(
  c: Context,
  cfg: PanelConfig,
  task: string,
): Promise<Response> {
  const version = c.req.query('version') ?? 'current'
  const subPath = c.req.query('path') ?? ''
  await limiter.acquire()
  try {
    const versionDir = await resolveVersionDir(cfg, task, version)
    const file = await resolveFileUnder(versionDir, subPath)
    const fi = await stat(file)
    const stream = Readable.toWeb(createReadStream(file)) as ReadableStream
    return c.body(stream, 200, {
      'Content-Type': 'application/octet-stream',
      'Content-Length': String(fi.size),
      'Content-Disposition': contentDisposition(basename(subPath)),
    })
  } finally {
    limiter.release()
  }
}

/** 版本预估体积：读版本 manifest 元数据；无则返回 null。 */
async function estimateVersionBytes(versionDir: string): Promise<number | null> {
  try {
    const raw = await import('node:fs/promises').then((m) => m.readFile(join(versionDir, MANIFEST_FILE), 'utf8'))
    const mf = JSON.parse(raw) as { entries?: { size?: number }[] }
    let sum = 0
    for (const e of mf.entries ?? []) sum += e.size ?? 0
    return sum
  } catch {
    return null
  }
}

const MB = 1024 * 1024

/** GET /api/tasks/:name/archive?version=&store=1 —— 整版本 zip 流式打包。 */
export async function handleArchive(
  c: Context,
  cfg: PanelConfig,
  task: string,
): Promise<Response> {
  const version = c.req.query('version') ?? 'current'
  const store = c.req.query('store') === '1'
  await limiter.acquire()
  try {
    const versionDir = await resolveVersionDir(cfg, task, version)
    const est = await estimateVersionBytes(versionDir)

    const archive = archiver('zip', { store })
    const pass = new PassThrough()
    archive.pipe(pass)
    // 目录整棵打包；元数据文件排除（对外不可见）。
    archive.directory(versionDir, false, (entry) =>
      basename(entry.name) === MANIFEST_FILE ? false : entry,
    )
    void archive.finalize()

    const headers: Record<string, string> = {
      'Content-Type': 'application/zip',
      'Content-Disposition': contentDisposition(`${task}-${version === 'current' ? 'current' : version}.zip`),
      'Cache-Control': 'no-store',
    }
    if (est !== null) {
      headers['X-Edge-Sync-Bytes'] = String(est)
      if (est > 500 * MB) headers['X-Edge-Sync-Suggest'] = 'rsync'
    }

    const body = Readable.toWeb(pass) as ReadableStream
    return c.body(body, 200, headers)
  } finally {
    limiter.release()
  }
}

File: panel/goserver/download.go
// 下载流与整版本 zip：fs 直读（不过内核），路径白名单 + 并发限流。
package main

import (
	"archive/zip"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

const manifestFile = ".edge-sync-manifest.json" // 与内核 engine.ManifestFile 对齐

var versionNameRe = regexp.MustCompile(`^\d{8}-\d{6}(-\d+)?$`)

const MB = 1024 * 1024

// downloadLimiter 全局并发上限 + 排队（超出 maxQueue 拒绝）。
type downloadLimiter struct {
	sem chan struct{}
	mu  struct {
		waiting int
	}
	maxQueue int
}

func newLimiter(max int) *downloadLimiter {
	return &downloadLimiter{sem: make(chan struct{}, max), maxQueue: 8}
}

func (l *downloadLimiter) acquire(w http.ResponseWriter) bool {
	if len(l.sem) >= cap(l.sem) && l.mu.waiting >= l.maxQueue {
		writeJSON(w, http.StatusTooManyRequests, map[string]string{"error": "too many concurrent downloads"})
		return false
	}
	select {
	case l.sem <- struct{}{}:
		return true
	default:
		l.mu.waiting++
		l.sem <- struct{}{}
		l.mu.waiting--
		return true
	}
}

func (l *downloadLimiter) release() {
	<-l.sem
}

// resolveVersionDir 把 "current" 或版本目录名解析为真实版本目录（防穿越）。
func resolveVersionDir(dataDir, task, version string) (string, error) {
	if task == "" || strings.ContainsAny(task, "/\\") || strings.Contains(task, "..") {
		return "", &httpError{status: http.StatusBadRequest, msg: "invalid task name"}
	}
	dataRoot := filepath.Join(dataDir, task)
	if version == "" || version == "current" {
		link := filepath.Join(dataRoot, "current")
		real, err := filepath.EvalSymlinks(link)
		if err != nil {
			return "", &httpError{status: http.StatusNotFound, msg: "no current version yet"}
		}
		return real, nil
	}
	if !versionNameRe.MatchString(version) {
		return "", &httpError{status: http.StatusBadRequest, msg: fmt.Sprintf("invalid version name: %s", version)}
	}
	dir := filepath.Join(dataRoot, "versions", version)
	fi, err := os.Stat(dir)
	if err != nil || !fi.IsDir() {
		return "", &httpError{status: http.StatusNotFound, msg: fmt.Sprintf("version not found: %s", version)}
	}
	return filepath.EvalSymlinks(dir)
}

// resolveFileUnder 版本目录内的文件解析：拒绝越界（SafeJoin 语义）、必须常规文件。
func resolveFileUnder(versionDir, subPath string) (string, error) {
	if subPath == "" || strings.Contains(subPath, "\x00") {
		return "", &httpError{status: http.StatusBadRequest, msg: "invalid path"}
	}
	clean := filepath.Clean(filepath.Join(versionDir, subPath))
	if !strings.HasPrefix(clean, versionDir+string(os.PathSeparator)) {
		return "", &httpError{status: http.StatusBadRequest, msg: "path escapes version directory"}
	}
	fi, err := os.Lstat(clean)
	if err != nil {
		return "", &httpError{status: http.StatusNotFound, msg: fmt.Sprintf("file not found: %s", subPath)}
	}
	if !fi.Mode().IsRegular() {
		return "", &httpError{status: http.StatusBadRequest, msg: "not a regular file"}
	}
	return clean, nil
}

// httpError 带 HTTP 状态的错误。
type httpError struct {
	status int
	msg    string
}

func (e *httpError) Error() string { return e.msg }

func (s *panelServer) handleDownload(w http.ResponseWriter, r *http.Request) {
	if !s.dl.acquire(w) {
		return
	}
	defer s.dl.release()

	versionDir, err := resolveVersionDir(s.cfg.DataDir, r.PathValue("name"), r.URL.Query().Get("version"))
	if err != nil {
		writeHErr(w, err)
		return
	}
	file, err := resolveFileUnder(versionDir, r.URL.Query().Get("path"))
	if err != nil {
		writeHErr(w, err)
		return
	}
	f, err := os.Open(file)
	if err != nil {
		writeHErr(w, err)
		return
	}
	defer f.Close()
	fi, _ := f.Stat()

	name := filepath.Base(r.URL.Query().Get("path"))
	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename*=UTF-8''%s", urlPathEscape(name)))
	// ServeContent：自带 Range 断点续传与 Last-Modified。
	http.ServeContent(w, r, name, fi.ModTime(), f)
}

// estimateVersionBytes 读版本 manifest 估算体积；无元数据返回 0。
func estimateVersionBytes(versionDir string) int64 {
	raw, err := os.ReadFile(filepath.Join(versionDir, manifestFile))
	if err != nil {
		return 0
	}
	var mf struct {
		Entries []struct {
			Size int64 `json:"size"`
		} `json:"entries"`
	}
	if json.Unmarshal(raw, &mf) != nil {
		return 0
	}
	var sum int64
	for _, e := range mf.Entries {
		sum += e.Size
	}
	return sum
}

func (s *panelServer) handleArchive(w http.ResponseWriter, r *http.Request) {
	if !s.dl.acquire(w) {
		return
	}
	defer s.dl.release()

	version := r.URL.Query().Get("version")
	if version == "" {
		version = "current"
	}
	store := r.URL.Query().Get("store") == "1"

	versionDir, err := resolveVersionDir(s.cfg.DataDir, r.PathValue("name"), version)
	if err != nil {
		writeHErr(w, err)
		return
	}

	task := r.PathValue("name")
	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition",
		fmt.Sprintf("attachment; filename*=UTF-8''%s", urlPathEscape(task+"-"+version+".zip")))
	w.Header().Set("Cache-Control", "no-store")
	if est := estimateVersionBytes(versionDir); est > 0 {
		w.Header().Set("X-Edge-Sync-Bytes", strconv.FormatInt(est, 10))
		if est > 500*MB {
			w.Header().Set("X-Edge-Sync-Suggest", "rsync")
		}
	}

	zipWriter := zip.NewWriter(w)
	defer zipWriter.Close()

	// 递归写入版本目录（元数据文件排除）。
	err = filepath.WalkDir(versionDir, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		if d.Name() == manifestFile {
			return nil
		}
		fi, err := d.Info()
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(versionDir, p)
		if err != nil {
			return err
		}
		hdr := &zip.FileHeader{
			Name:     filepath.ToSlash(rel),
			Modified: fi.ModTime(),
		}
		if store {
			hdr.Method = zip.Store
		}
		src, err := os.Open(p)
		if err != nil {
			return err
		}
		defer src.Close()
		dst, err := zipWriter.CreateHeader(hdr)
		if err != nil {
			return err
		}
		_, err = io.Copy(dst, src)
		return err
	})
	if err != nil && !errors.Is(err, fs.SkipDir) {
		// 头部已发送，无法改状态码；记录并中断。
		log.Printf("[panel] archive error: %v", err)
	}
}

// ---------- shared ----------

func urlPathEscape(name string) string {
	var b strings.Builder
	for _, r := range name {
		if r == '/' || r == '-' || r == '.' || r == '_' ||
			(r >= '0' && r <= '9') || (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') {
			b.WriteRune(r)
		} else {
			for _, by := range []byte(string(r)) {
				fmt.Fprintf(&b, "%%%02X", by)
			}
		}
	}
	return b.String()
}

func writeHErr(w http.ResponseWriter, err error) {
	var he *httpError
	if errors.As(err, &he) {
		writeJSON(w, he.status, map[string]string{"error": he.msg})
		return
	}
	writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
}
